libp2p/Peer.cpp

```diff
@@ -45,18 +45,33 @@ Peer::Peer(Peer const& _original):
 
 bool Peer::shouldReconnect() const
 {
-    return id && endpoint && chrono::system_clock::now() > m_lastAttempted + chrono::seconds(fallbackSeconds());
+    return id && endpoint &&
+           fallbackSeconds() != numeric_limits<unsigned>::max() &&
+           chrono::system_clock::now() > m_lastAttempted + chrono::seconds(fallbackSeconds());
 }
     
 unsigned Peer::fallbackSeconds() const
 {
     if (peerType == PeerType::Required)
         return 5;
+
+    switch (m_lastHandshakeFailure)
+    {
+        case FrameDecryptionFailure:
+        case ProtocolError:
+            return numeric_limits<unsigned>::max();
+        default:
+            break;
+    }
+
     switch (m_lastDisconnect)
     {
     case BadProtocol:
-        return 30 * (m_failedAttempts + 1);
     case UselessPeer:
+    case IncompatibleProtocol:
+    case UnexpectedIdentity:
+    case UserReason:
+        return numeric_limits<unsigned>::max();
     case TooManyPeers:
         return 25 * (m_failedAttempts + 1);
```

Comment:
`m_failedAttempts` seems to affect only the value of fallbackSeconds currently. Maybe we should use it now to retry several times and then go to "critical error, disconnect" state.
(at least for some cases of failures)

Reply:
@gumb0 Good idea - what do you think of this being taken care of in another PR? I'd like to limit the amount of changes I make to the peer gc logic in this PR so if something ends up breaking it will be easier to debug.

Reply:
Ok for another PR, but it seems to be the matter of only adding "m_failedAttempts >= 20" condition to `uselessPeer` function.

Reply:
Additional thought is that we could make all this change more conservative if we leave `fallbackSeconds()` as it was before (so that we do the same reconnects as before) and have just this check for failed attempts count in `uselessPeer` (plus the new check for handshake failures)
This way it would reconnect with the same intervals for each case as before, but stop after limited number of attempts.

But I'm fine with it if you think it's better to immediately stop in some cases.

Reply:
@gumb0 My hesitancy is that I don't know of a good value to use for the failed connection count threshold 😄 I'll ping the Geth and Parity folks to see if they have similar logic and if so, what their values are.